### PR TITLE
[5.2] Added SparkPost support

### PR DESCRIPTION
**Pull request by billmn**
I have added SparkPost support as email delivery service (like Mailgun and Mandrill).

SparkPost allows to send 100.000 emails/month for free.
For more informations: https://www.sparkpost.com/

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
How popular is it?

**Comment by arthurkirkosa** [comment]
It ranks "best" since Mandrill decided to go paid.

**Comment by billmn** [comment]
Currently I haven't statistics that allow me to tell how much this service is popular, but I can tell you that SparkPost was recommended from Mailchimp as a Mandril alternative: http://blog.mailchimp.com/important-changes-to-mandrill/

CTO Alec Peterson: promise to developers
https://www.sparkpost.com/blog/my-promise-to-developers-sparkpost-pricing

Mandrill migration guide:
https://www.sparkpost.com/mandrill-migration-guide

> SparkPost will honor Mandrill’s pricing for any customer that switches to SparkPost from Mandrill. Our plans include up to 100,000 free emails per month, which will be in effect for the life of the account, and includes incredibly cost-effective rates at higher message volumes.

**Comment by barryvdh** [comment]
Is there a benefit of using the API over SMTP?

**Comment by billmn** [comment]
The API drive return the JSON response from the remote webservice with information such as the message ID

**Comment by jonnywilliamson** [comment]
Thanks for this! (I know it isn't merged yet)

File: src/Illuminate/Mail/Transport/SparkPostTransport.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+     * @var string
+     */
+    protected $key;
+
+    /**
+     * Create a new SparkPost transport instance.
+     *
+     * @param  \GuzzleHttp\ClientInterface  $client
+     * @param  string  $key
+     * @param  string  $domain
```
Comment (helmut) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Noticed this extra param in docblock

File: src/Illuminate/Mail/Transport/SparkPostTransport.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * Get all the addresses this message should be sent to.
+     *
+     * Note that SparkPost still respects CC, BCC headers in raw message itself.
+     *
+     * @param  \Swift_Mime_Message $message
+     * @return array
+     */
+    protected function getRecipients(Swift_Mime_Message $message)
```
Comment (taylorotwell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: This looks to be copied from Mandrill. Does SparkPost actually work the same way or no?

File: src/Illuminate/Mail/Transport/SparkPostTransport.php (hunk 1 above)
Comment (billmn) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: You're right, I'll fix it

File: src/Illuminate/Mail/Transport/SparkPostTransport.php (hunk 2 above)
Comment (billmn) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Similar but not the same and yes, is copied from Mandrill and changed here: https://github.com/laravel/framework/pull/12579/files#diff-a92e9ec1ecc36811d8edb01394beb2f6R88

**Comment by taylorotwell** [comment]
So have you tested the driver with a SparkPost account to verify that it works?

**Comment by billmn** [comment]
Sure, and this is my tests :)

schermata 2016-03-04 alle 14 59 11

**Comment by tusharvikky** [comment]
Can we have sparkpost support for L5.1?

**Comment by GrahamCampbell** [comment]
> Can we have sparkpost support for L5.1?

NO.

**Comment by GrahamCampbell** [comment]
> Can we have sparkpost support for L5.1?

No, sorry. Laravel 5.1 only accepts bug fixes. Upgrade to 5.2 for new features.

File: src/Illuminate/Mail/Transport/SparkPostTransport.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function send(Swift_Mime_Message $message, &$failedRecipients = null)
+    {
+        $this->beforeSendPerformed($message);
+
+        $recipients = $this->getRecipients($message);
+
+        $message->setBcc([]);
```
Comment (GrahamCampbell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Why is this needed @billmn?

File: src/Illuminate/Mail/Transport/SparkPostTransport.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        $options = [
+            'headers' => [
+                'Authorization' => $this->key,
+            ],
+            'json' => [
```
Comment (GrahamCampbell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Are you sure all this will actually work properly on both guzzle 5 and 6?

File: src/Illuminate/Mail/Transport/SparkPostTransport.php (hunk 3 above)
Comment (billmn) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: To fix visibile recipients like Mailgun: https://github.com/laravel/framework/pull/10722

File: src/Illuminate/Mail/Transport/SparkPostTransport.php (hunk 4 above)
Comment (billmn) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Yes I have tested it

Comment (GrahamCampbell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: :+1:

File: src/Illuminate/Mail/Transport/SparkPostTransport.php (hunk 3 above)
Comment (GrahamCampbell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Ah, right, I see.

Comment (GrahamCampbell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Does this issue not affect mandrill also?

Comment (billmn) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: Honestly I don't remember ... Tomorrow morning I can test it if you would

Comment (GrahamCampbell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: That would be brilliant. Thank you. ;)

Comment (billmn) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: @GrahamCampbell I can confirm that this issue does not affect Mandrill driver.

Comment (GrahamCampbell) [inline src/Illuminate/Mail/Transport/SparkPostTransport.php]: :+1:
